### PR TITLE
Upload Dev Manifest Files to Azure Blob Storage

### PR DESCRIPTION
**Pull request by sonalideshpandemsft**
This PR uploads dev manifest files to azure blob. The following bunch of steps are added to the pipeline:

1. Run `flub release report -g client -o <path_generate_manifest_files> --baseFileName foo` - This command generates the manifest file as per the current state of the repository. It names the manifest files as `foo.full.json, foo.caret.json, foo.simple.json, foo.tilde.json`

2. Run `flub release report-unreleased -V <dev_version> -c <path_caret_manifest> -s <path_simple_manifest> -o <output_directory>` - This command accepts the dev version that is set in the pipeline. Later, it updates the generated manifest files in the pipeline pointing the values to the dev version. The manifest files undergo renaming in two formats: one based on the date and the other on the build number.

3. Later, after uploading the dev manifest files to azure blob, the manifest files are deleted to avoid failures at the step `Check for extraneous modified files`.

Pipeline Run: https://dev.azure.com/fluidframework/internal/_build/results?buildId=259575&view=logs&s=6884a131-87da-5381-61f3-d7acc3b91d76

This is how the upload folder looks:
image

[AB#7884](https://dev.azure.com/fluidframework/internal/_workitems/edit/7884)

### CODE REVIEW

**Review by scottn12 — APPROVED** [review]
LGTM! Not sure if you want explicit approval from a build-tools owner but I don't see any issues on my end

**Comment by sonalideshpandemsft** [comment]
Thanks @scottn12!! not sure the why the CI checks aren't kicking in for this PR

File: tools/pipelines/templates/upload-dev-manifest.yml
Comment (tylerbutler) [inline tools/pipelines/templates/upload-dev-manifest.yml:57]: This also gets easier if you output everything to a folder. Maybe add an --outDir flag to the command?

File: tools/pipelines/templates/upload-dev-manifest.yml
Comment (tylerbutler) [inline tools/pipelines/templates/upload-dev-manifest.yml:16]:
```suggestion
  displayName: Generate release reports
```

File: tools/pipelines/templates/upload-dev-manifest.yml
Comment (tylerbutler) [inline tools/pipelines/templates/upload-dev-manifest.yml:24]:
```suggestion
  displayName: Update release report version
```

File: tools/pipelines/templates/upload-dev-manifest.yml
Comment (tylerbutler) [inline tools/pipelines/templates/upload-dev-manifest.yml:35]:
```suggestion
  displayName: Upload release reports
```

**Review by tylerbutler — CHANGES_REQUESTED** [review]

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/release/report-unreleased.ts:88]: The `report-unreleased` command generates caret and simple manifest files. 

The function `isInternalVersionRange` returns true if a string represents an internal version range, replacing the versions correctly in caret manifest file but the simple manifest file points to exact version range. The second condition is to replace simple manifest file versions.

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/release/report-unreleased.ts:15]: The command just generates the report. Uploading is done as another task in the pipeline.

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/release/report-unreleased.ts:94]: toISOString() returns a string including date, time, and timezone. By using .slice(0, 10), I'm extracting only the date part, keeping the first 10 characters. The date is needed without the time and timezone information.

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/release/report-unreleased.ts]: Replaced the second condition with the function `isInternalVersionScheme` which checks if a version matches the Fluid internal version scheme.

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/release/report-unreleased.ts:34]:
```suggestion
		caretManifestFilePath: Flags.file({
			description: "Path to a report file in the 'caret' format.",
			exists: true,
		}),
```

Use the file flag and make oclif check the file exists for you.

You may need to update the release report command so that you can output specific file names. Basically avoid searching for the files you need - rather make it so the file paths are always provided - which _might_ require changing release report to optionally output names without variables like the version.

Maybe release report could take a --baseFileName flag and if provided it outputs the files with names like `baseFileName.caret.json`?

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/release/report-unreleased.ts:39]:
```suggestion
		simpleManifestFilePath: Flags.file({
			description: "Path to a report file in the 'simple' format.",
			exists: true,
		}),
```

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/release/report-unreleased.ts:97]: I've been thinking about this a bit, and the `full` report version has everything that the other reports do - why not take the full report in as input and transform it as needed? You can read in the JSON file cast it as a ReleaseReport type and then transform as needed. See toReportKind in library/release.ts too.

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/release/report-unreleased.ts:125]: Why do you need the void return? Can you always return a string|undefined?

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/release/report-unreleased.ts:135]: Suggest using [date-fns](https://date-fns.org/) to deal with this. It will be clearer and more self-documenting. We already have the dependency I think.

File: build-tools/packages/build-cli/src/commands/release/report-unreleased.ts
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/release/report-unreleased.ts]: That's a good idea. I have pushed the changes, but code cleanup is remaining. I'll work on that in sometime.

File: build-tools/packages/build-cli/src/commands/release/report.ts
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/release/report.ts:389]: Make this a string and then use it as the base filename. E.g. `--baseFileName foo` would yield `foo.caret.json`.

**Comment by sonalideshpandemsft** [comment]
hey @tylerbutler - any ideas why the CI does not run for this PR? I tried to revert `build-npm-package.yml` and `include-set-package-version.yml` changes: https://github.com/microsoft/FluidFramework/pull/20948 and the CI jobs kick in. Should the changes be staged in some way?

**Review by tylerbutler — COMMENTED** [review]
Biggest feedback is to make all the function work with ReleaseReport objects, update the versions there, then convert to whatever other formats you need just before writing the files.

**Review by tylerbutler — APPROVED** [review]
You can merge as-is but there are a couple of minor things. Fine to follow up in a separate PR if CI is passing though.

**Pull request merged**
